Add load reporter for server load reporting

AspirinSJL opened the pull request:
This is a continuation of #14793 ~and currently based on that PR~.

~This is not yet ready for review. I hope to see some quick feedback from @isturdy on my usage of the opencensus APIs.~

This is ready for review except for the `get_cpu_stats()` implementation on the platforms other than Linux.

I skipped some low-risk but detailed tests for now.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/15196)
<!-- Reviewable:end -->

AspirinSJL commented on src/cpp/server/load_reporter/load_reporter.cc:
```diff
@@ -0,0 +1,396 @@
+  return get_cpu_stats();
+}
+
+CensusViewProviderDefaultImpl::CensusViewProviderDefaultImpl() {
```
@isturdy This is how I create the views.

AspirinSJL commented on src/cpp/server/load_reporter/load_reporter.cc:
```diff
@@ -0,0 +1,396 @@
+  load_data_store_.ReportStreamClosed(hostname, lb_id);
+}
+
+void LoadReporter::FetchAndSample() {
```
@isturdy This is how I process the ViewData.

AspirinSJL commented on src/cpp/server/load_reporter/load_reporter.cc:
```diff
@@ -0,0 +1,396 @@
+  view_map_.insert({kViewOtherCallMetricValue, View(vd_metric_value)});
+}
+
+CensusViewProvider::ViewDataMap CensusViewProviderDefaultImpl::FetchViewData() {
```
@isturdy This is how I fetch the data.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

isturdy commented on src/cpp/server/load_reporter/load_reporter.h:
```diff
@@ -0,0 +1,169 @@
+#include "src/cpp/server/load_reporter/load_data_store.h"
+#include "src/proto/grpc/lb/v1/load_reporter.grpc.pb.h"
+
+#include "third_party/opencensus-cpp/opencensus/stats/aggregation.h"
```
You should be able to replace all these imports with "third_party/opencensus-cpp/opencensus/stats/stats.h"

AspirinSJL replied:
Got it. Thanks!

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

AspirinSJL commented on WORKSPACE:
```diff
@@ -3,3 +3,16 @@ workspace(name = "com_github_grpc_grpc")
 grpc_deps()
 grpc_test_only_deps()
+
+http_archive(
```
@nicolasnoble Is the change in this file OK?

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented:
Something is wrong with my merging from upstream. I will clean it tomorrow.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented:
Thanks for reviewing!

---

Review status: 19 of 23 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[build.yaml, line 1894 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQNLw6L8duT3SQji1:-LDgoZgq3cSUJm-kRUsa:b1j3mgf) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/build.yaml#L1894)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I thought we decided that this won't build via make?
</blockquote></details>

I keep this for now because it seems that we still need some discussion about this? See https://github.com/grpc/grpc/pull/15349#discussion_r188754176

---

*[src/cpp/server/load_reporter/get_cpu_stats.h, line 29 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgR0CO71OleigDA8rA:-LDgp_ufD_Q_g3kuuF5H:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/get_cpu_stats.h#L29)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This function name should be C++-style instead of C style: `GetCpuStats()`.
</blockquote></details>

Done.

---

*[src/cpp/server/load_reporter/load_data_store.h, line 78 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRluq3mEZ4GYXX_6q:-LDgr5qwEZnhfQXm44uj:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L78)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

No need for `explicit` here, since there are 2 parameters.

Actually, it looks like we're unnecessarily using `explicit` even for some of the existing ctors.  Please remove throughout, except in cases where the ctor has a single parameter.
</blockquote></details>

Done.

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LDgrB3Q1-T5-lhUhTzC:begsem3) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Why are these parameters changing from int64 to double?

For that matter, why are some of the parameters doubles even before this PR?  Shouldn't these all be integers?  It's not clear to me why we would have fractional values for any of these.
</blockquote></details>

Logically they should be integers. But there is some tricky problem with OpenCensus here: when making testing view data, OpenCensus will treat the input values as separate measurements and add them together according to the data type (i.e., if it's int, add 1; if it's double, add the original input value). For the testing to be effective and the code length to be reasonable, I changed the data type to double.

But after upon thought, I should actually make a feature request or maybe a pull request to OpenCensus. @isturdy Does it sound reasonable to add a method to set the final view data values directly?

---

*[src/cpp/server/load_reporter/load_data_store.cc, line 81 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgTC_--TOe7oTumi5c:-LDgx-N18C9wEuSPo2tn:b62d3t2) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.cc#L81)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I would avoid using `std::stoi()`, since it can generate exceptions, which should not be used in Google C++ code, as per the style guide.

Since this code imposes a transitive dependency on absl anyway, consider using `absl::SimpleAtoi()` instead.
</blockquote></details>

Oh, good to know that.

Changed to `sscanf()` to be consistent with a change below.

---

*[src/cpp/server/load_reporter/load_data_store.cc, line 99 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgVkPz8Q4-AZBaHo9x:-LDh1MWo9bKOAmdZWUF4:b-ohjkaf) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.cc#L99)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Similarly, suggest using `absl::SimpleAtoi()` instead of `std::stoul()`.  Same thing throughout this function.
</blockquote></details>

Done.

Since `absl::SimpleAtoi()` only supports base of 10, I changed to `sscanf()`

---

*[src/cpp/server/load_reporter/load_reporter.h, line 67 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgXaBe9foQGa6xhg6m:-LDh45tl7ecLnhunvhne:btdsfzr) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_reporter.h#L67)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Data members should always be private, not protected.

https://google.github.io/styleguide/cppguide.html#Access_Control
</blockquote></details>

I see.

---

*[src/cpp/server/load_reporter/load_reporter.cc, line 212 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgZ4OY81qjKYDhhsNn:-LDhWx1v2wUumH23Nyj0:b-scttgq) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_reporter.cc#L212)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Consider using `absl::StrFormat()`.
</blockquote></details>

Unfortunately, I can't find str_format.h in our submodule. I guess it's a version problem?

Anyways, I changed to `sprintf()`, which seems similar to `absl::StrFormat()`.

---

*[src/cpp/server/load_reporter/util.h, line 25 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgSdSw97U3AImOop_6:-LDhXr3vDMuZHWUZVnqS:bs7wumb) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/util.h#L25)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Since all that's in this file is constants, how about renaming it from util.h to constants.h?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196#-:-LDgo2QPAfMXrhXNSUh1:b-yz40g0)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

isturdy commented:
Review status: 7 of 23 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LDhh32K3m24FoOrsxr3:b4v67kk) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

Logically they should be integers. But there is some tricky problem with OpenCensus here: when making testing view data, OpenCensus will treat the input values as separate measurements and add them together according to the data type (i.e., if it's int, add 1; if it's double, add the original input value). For the testing to be effective and the code length to be reasonable, I changed the data type to double.

But after upon thought, I should actually make a feature request or maybe a pull request to OpenCensus. @isturdy Does it sound reasonable to add a method to set the final view data values directly?
</blockquote></details>

The only public interface for making ViewData for testing is in https://github.com/census-instrumentation/opencensus-cpp/blob/master/opencensus/stats/testing/test_utils.h. (Otherwise, the only way to populate a ViewData without using implementation headers should be the normal stats workflow: make a View, record data, and then get data for the view.)

We would rather not expose a means of setting final data directly (i.e. bypassing aggregation) to avoid exposing implementation details.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented on src/cpp/server/load_reporter/util.h:
```diff
@@ -0,0 +1,106 @@
+// Measures.
+
+::opencensus::stats::MeasureInt64 GetMeasureStartCount() {
+  static const ::opencensus::stats::MeasureInt64 start_count =
```
According to the style guide https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables this static variable is forbidden. But this is the recommended style from the OpenCensus [example](https://github.com/census-instrumentation/opencensus-cpp/blob/13e95557eaa03fdcfbafb1a005298b03353ff34d/opencensus/stats/examples/view_and_recording_example.cc#L30). 

@isturdy  So I'm wondering if there is any recommended way to get the measures and the tags without violating the style guide.

isturdy replied:
This has local scope and Measure* and TagKey have trivial destructors, so the style guide does allow static storage.

AspirinSJL replied:
That sounds great. Thanks!

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented:
Reviewed 3 of 11 files at r2, 12 of 13 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LEFfnYj1FZToQqj1u9i:b-btwnpt) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, isturdy (Ian Sturdy) wrote…</i></summary><blockquote>

The only public interface for making ViewData for testing is in https://github.com/census-instrumentation/opencensus-cpp/blob/master/opencensus/stats/testing/test_utils.h. (Otherwise, the only way to populate a ViewData without using implementation headers should be the normal stats workflow: make a View, record data, and then get data for the view.)

We would rather not expose a means of setting final data directly (i.e. bypassing aggregation) to avoid exposing implementation details.
</blockquote></details>

Even if the opencensus API requires doubles, I don't see why we need to use that type in our code.  We should use int64 here and convert to/from double at the points at which we interact with opencensus.

---

*[src/cpp/server/load_reporter/load_data_store.cc, line 83 at r6](https://reviewable.io/reviews/grpc/grpc/15196#-LEFgHTD2LJGlPdh8-CT:-LEFgHTD2LJGlPdh8-CU:b992jt6) ([raw file](https://github.com/grpc/grpc/blob/1cfafc69a8525378b4e8a7e322fc30e01e17f699/src/cpp/server/load_reporter/load_data_store.cc#L83)):*
> ```cpp
>   GPR_ASSERT(client_ip_and_token.size() >= 2);
>   int ip_hex_size;
>   sscanf(client_ip_and_token.substr(0, 2).c_str(), "%d", &ip_hex_size);
> ```

I think we need to check the return value here to ensure that sscanf() actually processed one item.

Same for all calls to sscanf() below.

---

*[src/cpp/server/load_reporter/load_reporter.cc, line 211 at r6](https://reviewable.io/reviews/grpc/grpc/15196#-LEFguqP0zD8bovl33db:-LEFguqQ2V_bpceY_y2X:b33994d) ([raw file](https://github.com/grpc/grpc/blob/1cfafc69a8525378b4e8a7e322fc30e01e17f699/src/cpp/server/load_reporter/load_reporter.cc#L211)):*
> ```cpp
>     // Convert to padded hex string for a 32-bit LB ID. E.g, "0000ca5b".
>     char buf[kLbIdLength];
>     sprintf(buf, "%08lx", lb_id);
> ```

Suggest using `snprintf()` for safety.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

AspirinSJL commented:
Review status: all files reviewed, 4 unresolved discussions (waiting on @markdroth, @isturdy, @summerxyt, and @nicolasnoble)

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LEjg5HT-HL-uocQX1XZ:bboepgj) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Even if the opencensus API requires doubles, I don't see why we need to use that type in our code.  We should use int64 here and convert to/from double at the points at which we interact with opencensus.
</blockquote></details>

I see. Done.

---

*[src/cpp/server/load_reporter/load_data_store.cc, line 83 at r6](https://reviewable.io/reviews/grpc/grpc/15196#-LEFgHTD2LJGlPdh8-CT:-LEjlBkHEfuIm-011sbG:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/1cfafc69a8525378b4e8a7e322fc30e01e17f699/src/cpp/server/load_reporter/load_data_store.cc#L83)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think we need to check the return value here to ensure that sscanf() actually processed one item.

Same for all calls to sscanf() below.
</blockquote></details>

Done.

---

*[src/cpp/server/load_reporter/load_reporter.cc, line 211 at r6](https://reviewable.io/reviews/grpc/grpc/15196#-LEFguqP0zD8bovl33db:-LEjpAkE7T6hvy47go_K:b-qeks77) ([raw file](https://github.com/grpc/grpc/blob/1cfafc69a8525378b4e8a7e322fc30e01e17f699/src/cpp/server/load_reporter/load_reporter.cc#L211)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Suggest using `snprintf()` for safety.
</blockquote></details>

Done.

I didn't realize that `sprintf()` is null-terminated. Thanks!

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented:
This looks pretty good -- just one issue left.

---

Reviewed 2 of 4 files at r7, 1 of 1 files at r8, 2 of 2 files at r9.
Review status: all files reviewed, 2 unresolved discussions (waiting on @markdroth, @isturdy, @summerxyt, and @nicolasnoble)

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LEuTdSO4hDRKlLzzf5J:b-3w9g1v) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

I see. Done.
</blockquote></details>

Can we also change the other args that use doubles to use ints?

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196#-:-LEuUW0x5gLg0W7f2glg:b-pbet80)*
<!-- Sent from Reviewable.io -->

yetianx commented on src/cpp/server/load_reporter/load_data_store.cc:
```diff
@@ -73,6 +75,54 @@ const typename C::value_type* RandomElement(const C& container) {
+    return "";
+  } else if (client_ip_hex_.size() == kIpv4AddressLength) {
+    uint32_t ip_bytes;
+    GPR_ASSERT(sscanf(client_ip_hex_.c_str(), "%x", &ip_bytes) == 1);
```
Does this ASSERT terminate the server?

It might be better to log an warning or error instead of crashing the server.

yetianx commented on src/cpp/server/load_reporter/load_data_store.h:
```diff
@@ -32,7 +32,6 @@ namespace grpc {
 namespace load_reporter {
 
 constexpr char kInvalidLbId[] = "<INVALID_LBID_238dsb234890rb>";
```
Also move constant.h?

yetianx commented on src/cpp/server/load_reporter/load_reporter.cc:
```diff
@@ -0,0 +1,484 @@
+    cpu_stats = {0, 0};
+  }
+  std::unique_lock<std::mutex> lock(feedback_mu_);
+  feedback_records_.push_back(
```
Use emplace_back?

yetianx commented on src/cpp/server/load_reporter/load_reporter.cc:
```diff
@@ -0,0 +1,484 @@
+          hostname.c_str(), lb_id.c_str());
+}
+
+void LoadReporter::FetchAndSample() {
```
nit: It might be better to split this function into 4 pieces as your comment suggests. And you have a helper function for each piece. Then each helper function is within 50 lines.

AspirinSJL commented:
Thanks for reviewing!

---

Review status: all files reviewed, 5 unresolved discussions (waiting on @markdroth, @isturdy, @yetianx, @summerxyt, and @nicolasnoble)

---

*[src/cpp/server/load_reporter/load_data_store.h, line 127 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgRvdn7d9qohcv6Km1:-LDhxXBOBO9ebyEcokAB:b-o3ygbo) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/src/cpp/server/load_reporter/load_data_store.h#L127)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Can we also change the other args that use doubles to use ints?
</blockquote></details>

Done.

Theses are all ints from the channel filter.

---

*[src/cpp/server/load_reporter/load_data_store.h, line 34 at r9](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQ8Cx5_eXG6s76-Vz-r9-34:-LFIs8EX4Xg-HFoCj9Hm:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/134e40b0821287935c0d89eb12938efcdacee7d6/src/cpp/server/load_reporter/load_data_store.h#L34)):*
<details><summary><i>Previously, yetianx wrote…</i></summary><blockquote>

Also move constant.h?
</blockquote></details>

Done.

---

*[src/cpp/server/load_reporter/load_data_store.cc, line 104 at r9](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQ8Cx5_eXG6s76-Vy-r9-104:-LFIiZrj0EJ7li61FXEi:b-slxgf9) ([raw file](https://github.com/grpc/grpc/blob/134e40b0821287935c0d89eb12938efcdacee7d6/src/cpp/server/load_reporter/load_data_store.cc#L104)):*
<details><summary><i>Previously, yetianx wrote…</i></summary><blockquote>

Does this ASSERT terminate the server?

It might be better to log an warning or error instead of crashing the server.
</blockquote></details>

Done.

Also, apparently I had some misunderstanding about the endianness here. I just added back the `hton` conversion.

---

*[src/cpp/server/load_reporter/load_reporter.cc, line 356 at r9](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQ8Cx5_eXG6s76-W--r9-356:-LFIkQp81j6iIrdw-gpP:b-ugvulk) ([raw file](https://github.com/grpc/grpc/blob/134e40b0821287935c0d89eb12938efcdacee7d6/src/cpp/server/load_reporter/load_reporter.cc#L356)):*
<details><summary><i>Previously, yetianx wrote…</i></summary><blockquote>

Use emplace_back?
</blockquote></details>

Done.

I was too lazy to add ctor to the struct.

---

*[src/cpp/server/load_reporter/load_reporter.cc, line 379 at r9](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQ8Cx5_eXG6s76-W--r9-379:-LFIpfJe92EPSwSzRQji:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/134e40b0821287935c0d89eb12938efcdacee7d6/src/cpp/server/load_reporter/load_reporter.cc#L379)):*
<details><summary><i>Previously, yetianx wrote…</i></summary><blockquote>

nit: It might be better to split this function into 4 pieces as your comment suggests. And you have a helper function for each piece. Then each helper function is within 50 lines.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196#-:-LFIpiy6-dmIQjNKp6_T:b-irrk52)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented:
This now builds on open PR #15799.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%    +128 [None]                                                  +432  +0.0%
  +0.9%     +32 src/core/lib/iomgr/socket_utils_common_posix.cc          +32  +0.9%
       +13%     +22 [Unmapped]                                               +22   +13%
      [NEW]      +7 grpc_ntohs(unsigned short) [clone .localalias.3]          +7  [NEW]
      [NEW]      +5 grpc_htonl                                                +5  [NEW]
      [NEW]      +5 grpc_ntohl                                                +5  [NEW]

  +0.0%    +160 TOTAL                                                   +464  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%    +128 [None]                                                  +432  +0.0%
  +0.9%     +32 src/core/lib/iomgr/socket_utils_common_posix.cc          +32  +0.9%
       +13%     +22 [Unmapped]                                               +22   +13%
      [NEW]      +7 grpc_ntohs(unsigned short) [clone .localalias.3]          +7  [NEW]
      [NEW]      +5 grpc_htonl                                                +5  [NEW]
      [NEW]      +5 grpc_ntohl                                                +5  [NEW]

  +0.0%    +160 TOTAL                                                   +464  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%    +128 [None]                                                  +432  +0.0%
  +0.9%     +32 src/core/lib/iomgr/socket_utils_common_posix.cc          +32  +0.9%
       +13%     +22 [Unmapped]                                               +22   +13%
      [NEW]      +7 grpc_ntohs(unsigned short) [clone .localalias.3]          +7  [NEW]
      [NEW]      +5 grpc_htonl                                                +5  [NEW]
      [NEW]      +5 grpc_ntohl                                                +5  [NEW]

  +0.0%    +160 TOTAL                                                   +464  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%    +128 [None]                                                  +432  +0.0%
  +0.9%     +32 src/core/lib/iomgr/socket_utils_common_posix.cc          +32  +0.9%
       +13%     +22 [Unmapped]                                               +22   +13%
      [NEW]      +7 grpc_ntohs(unsigned short) [clone .localalias.3]          +7  [NEW]
      [NEW]      +5 grpc_htonl                                                +5  [NEW]
      [NEW]      +5 grpc_ntohl                                                +5  [NEW]

  +0.0%    +160 TOTAL                                                   +464  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented:
All the failures seen have been fixed. This is ready for review again.

markdroth commented:
Other than the outstanding question for Nico about build/test integration, this looks good!

---

Reviewed 2 of 6 files at r10, 3 of 4 files at r11, 3 of 10 files at r12, 7 of 7 files at r13, 6 of 6 files at r14.
Review status: all files reviewed, 4 unresolved discussions (waiting on @markdroth, @yetianx, @summerxyt, and @nicolasnoble)

---

*[build.yaml, line 1894 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQNLw6L8duT3SQji1:-LFSs3lt9Jbr22vTKYVn:b4wbugo) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/build.yaml#L1894)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

I keep this for now because it seems that we still need some discussion about this? See https://github.com/grpc/grpc/pull/15349#discussion_r188754176
</blockquote></details>

Still need feedback from @nicolasnoble about this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196#-:-LFSsEdw1H9IXiRF1Bz-:b13fgzv)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

AspirinSJL commented:
Cool! I've cleaned it up. This PR is ready for a final review now. @yetianx @markroth.

---

Review status: 17 of 18 files reviewed, 4 unresolved discussions (waiting on @markdroth, @yetianx, @summerxyt, and @nicolasnoble)

---

*[build.yaml, line 1894 at r1](https://reviewable.io/reviews/grpc/grpc/15196#-LDgQNLw6L8duT3SQji1:-LFYfUU15tgDKu_R2u2v:b-2pnz9x) ([raw file](https://github.com/grpc/grpc/blob/68342ca6d74da6c3b10427fb7b1f32dcad3830f3/build.yaml#L1894)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Still need feedback from @nicolasnoble about this.
</blockquote></details>

Done. @nicolasnoble says the Bazel tests are running well. So I have removed these targets here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15196#-:-LFYezU40xSf54cW8ldd:b-9km3x)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
[trickle] No significant performance differences
```

AspirinSJL commented:
Squashing. (Heads-up to @yetianx)

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

AspirinSJL commented:
I have one small change between approval and squashing. 

We already have OpenCensus dependency [set up](https://github.com/grpc/grpc/pull/15070/files#diff-5952172e3b803ed640371cda4cddf13dR85) after #15070. So I removed my change to `WORKSPACE` and followed #15070 to add an `external_deps`.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

AspirinSJL commented:
Basic Tests MacOS [opt] (internal CI): #15027
